File: api/sdks/ruby/lib/objstore/clients/mcp_client.rb
require "net/http"
require "json"
require "base64"
require "openssl"

module ObjectStore
  module Clients
    # MCP (Model Context Protocol) HTTP client for ObjectStore.
    #
    # Communicates via HTTP POST JSON-RPC 2.0 to the MCP server endpoint at "/".
    # Operations are invoked with method "tools/call" and tool names prefixed
    # with "objstore_". Text results are returned in result.content[0].text.
    #
    # Streaming (put_stream/get_stream) is supported by buffering the IO in
    # memory, which is appropriate for request/response HTTP transports.
    #
    # @api private
    class McpClient
      include JsonRpcHelpers

      attr_reader :host, :port, :use_ssl, :timeout

      # Initialize a new MCP client.
      #
      # @param host [String] Server hostname
      # @param port [Integer] Server port
      # @param use_ssl [Boolean] Whether to use HTTPS
      # @param timeout [Integer] Request timeout in seconds
      # @param token [String, nil] Bearer token for Authorization header
      # @param headers [Hash] Additional HTTP headers to send on every request
      # @param tenant_id [String, nil] Tenant identifier sent as X-Tenant-ID header
      def initialize(host: "localhost", port: 8081, use_ssl: false, timeout: 30,
                     token: nil, headers: {}, tenant_id: nil)
        @host = host
        @port = port
        @use_ssl = use_ssl
        @timeout = timeout
        @token = token
        @extra_headers = headers || {}
        @tenant_id = tenant_id
        @id_counter = 0
        @id_mutex = Mutex.new
        @http = build_http
        @http_mutex = Mutex.new
      end

      # Upload an object to the store.
      #
      # @param key [String] Object key
      # @param data [String] Object data (may be binary)
      # @param metadata [Models::Metadata, Hash, nil] Optional metadata
      #
      # @return [Models::PutResponse] Response with success status
      #
      # @raise [ValidationError, ServerError, ConnectionError]
      def put(key, data, metadata = nil)
        metadata_obj = metadata.is_a?(Models::Metadata) ? metadata : Models::Metadata.new(metadata || {})
        # Object data travels base64-encoded so the transport is binary-safe.
        args = { "key" => key, "data" => Base64.strict_encode64(data.to_s) }
        args["metadata"] = build_metadata_args(metadata_obj) unless metadata_obj_empty?(metadata_obj)

        call_tool("objstore_put", args) do |result|
          Models::PutResponse.new(
            success: result.fetch("success", true),
            message: result["message"]
          )
        end
      end

      # Upload an object from an IO stream.
      #
      # Reads the entire IO into memory before sending (buffered).
      #
      # @param key [String] Object key
      # @param io [IO, #read] IO object to read from
      # @param metadata [Models::Metadata, Hash, nil] Optional metadata
      # @param chunk_size [Integer] Size of chunks to read
      #
      # @return [Models::PutResponse] Response with success status
      def put_stream(key, io, metadata: nil, chunk_size: 8192)
        data = String.new(encoding: "BINARY")
        while (chunk = io.read(chunk_size))
          data << chunk
        end
        put(key, data, metadata)
      end

      # Retrieve an object from the store.
      #
      # @param key [String] Object key
      #
      # @return [Models::GetResponse] Response with data and metadata
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def get(key)
        call_tool("objstore_get", { "key" => key }) do |result|
          Models::GetResponse.new(decode_object_data(result), Models::Metadata.new)
        end
      end

      # Retrieve an object and yield it as a single chunk.
      #
      # @param key [String] Object key
      #
      # @yieldparam chunk [String] The full object data as one chunk
      #
      # @return [Models::Metadata] Object metadata
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def get_stream(key)
        call_tool("objstore_get", { "key" => key }) do |result|
          data = decode_object_data(result)
          yield data if block_given?
          Models::Metadata.new
        end
      end

      # Delete an object from the store.
      #
      # @param key [String] Object key
      #
      # @return [Models::DeleteResponse] Response with success status
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def delete(key)
        call_tool("objstore_delete", { "key" => key }) do |result|
          Models::DeleteResponse.new(success: result.fetch("success", true), message: result["message"])
        end
      end

      # List objects in the store.
      #
      # @param prefix [String, nil] Filter by key prefix
      # @param delimiter [String, nil] Ignored; MCP tool does not expose delimiter
      # @param max_results [Integer] Maximum results to return
      # @param continue_from [String, nil] Pagination token from previous response
      #
      # @return [Models::ListResponse] List response with objects
      #
      # @raise [ServerError, ConnectionError]
      def list(prefix: nil, delimiter: nil, max_results: 100, continue_from: nil)
        args = {}
        args["prefix"] = prefix if prefix
        args["max_results"] = max_results if max_results
        args["continue_from"] = continue_from if continue_from

        call_tool("objstore_list", args) do |result|
          keys = result["keys"] || []
          objects = keys.map { |k| Models::ObjectInfo.new(key: k) }

          Models::ListResponse.new(
            objects: objects,
            next_token: result["next_token"],
            truncated: result.fetch("truncated", false)
          )
        end
      end

      # Check if an object exists.
      #
      # @param key [String] Object key
      #
      # @return [Boolean] true if the object exists
      #
      # @raise [ServerError, ConnectionError]
      def exists?(key)
        call_tool("objstore_exists", { "key" => key }) do |result|
          result.fetch("exists", false)
        end
      end

      # Get metadata for an object.
      #
      # @param key [String] Object key
      #
      # @return [Models::MetadataResponse] Response with metadata
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def get_metadata(key)
        call_tool("objstore_get_metadata", { "key" => key }) do |result|
          Models::MetadataResponse.new(
            metadata: Models::Metadata.new(
              content_type: result["content_type"],
              content_encoding: result["content_encoding"],
              size: result["size"],
              etag: result["etag"],
              last_modified: result["last_modified"],
              custom: result["custom"] || {}
            ),
            success: true
          )
        end
      end

      # Update metadata for an object.
      #
      # @param key [String] Object key
      # @param metadata [Models::Metadata, Hash] New metadata
      #
      # @return [Models::UpdateMetadataResponse] Response with success status
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def update_metadata(key, metadata)
        metadata_obj = metadata.is_a?(Models::Metadata) ? metadata : Models::Metadata.new(metadata)
        args = { "key" => key, "metadata" => build_metadata_args(metadata_obj) }

        call_tool("objstore_update_metadata", args) do |result|
          Models::UpdateMetadataResponse.new(success: result.fetch("success", true), message: result["message"])
        end
      end

      # Check server health status.
      #
      # @param service [String, nil] Ignored; included for API parity
      #
      # @return [Models::HealthResponse] Response with health status
      #
      # @raise [ServerError, ConnectionError]
      def health(service: nil)
        call_tool("objstore_health", {}) do |result|
          Models::HealthResponse.new(
            status: result.fetch("status", "healthy"),
            message: result["version"]
          )
        end
      end

      # Archive an object to external storage.
      #
      # @param key [String] Object key
      # @param destination_type [String] Archive destination type (e.g. "glacier")
      # @param destination_settings [Hash] Settings for the destination
      #
      # @return [Models::ArchiveResponse] Response with success status
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def archive(key, destination_type:, destination_settings: {})
        args = {
          "key" => key,
          "destination_type" => destination_type,
          "destination_settings" => destination_settings
        }

        call_tool("objstore_archive", args) do |result|
          Models::ArchiveResponse.new(success: result.fetch("success", true), message: result["message"])
        end
      end

      # Add a lifecycle policy.
      #
      # @param policy [Models::LifecyclePolicy, Hash] Policy to add
      #
      # @return [Hash] Response with :success and :message keys
      #
      # @raise [ValidationError, ServerError, ConnectionError]
      def add_policy(policy)
        policy_obj = policy.is_a?(Models::LifecyclePolicy) ? policy : Models::LifecyclePolicy.new(policy)
        args = {
          "id" => policy_obj.id,
          "prefix" => policy_obj.prefix || "",
          "retention_seconds" => policy_obj.retention_seconds || 0,
          "action" => policy_obj.action
        }
        args["destination_type"] = policy_obj.destination_type if policy_obj.destination_type
        args["destination_settings"] = policy_obj.destination_settings if policy_obj.destination_settings&.any?

        call_tool("objstore_add_policy", args) do |result|
          { success: result.fetch("success", true), message: result["message"] }
        end
      end

      # Remove a lifecycle policy.
      #
      # @param id [String] Policy ID
      #
      # @return [Hash] Response with :success and :message keys
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def remove_policy(id)
        call_tool("objstore_remove_policy", { "id" => id }) do |result|
          { success: result.fetch("success", true), message: result["message"] }
        end
      end

      # Get all lifecycle policies.
      #
      # @param prefix [String, nil] Filter by prefix
      #
      # @return [Hash] Response with :policies array and :success key
      #
      # @raise [ServerError, ConnectionError]
      def get_policies(prefix: nil)
        args = prefix ? { "prefix" => prefix } : {}

        call_tool("objstore_get_policies", args) do |result|
          policies = extract_policies(result).map { |p| parse_lifecycle_policy(p) }
          { policies: policies, success: true }
        end
      end

      # Apply all lifecycle policies.
      #
      # @return [Hash] Response with :success, :policies_count, :objects_processed, :message
      #
      # @raise [ServerError, ConnectionError]
      def apply_policies
        call_tool("objstore_apply_policies", {}) do |result|
          {
            success: result.fetch("success", true),
            policies_count: result["policies_count"] || 0,
            objects_processed: result["objects_processed"] || 0,
            message: result["message"]
          }
        end
      end

      # Add a replication policy.
      #
      # @param policy [Models::ReplicationPolicy, Hash] Policy to add
      #
      # @return [Hash] Response with :success and :message keys
      #
      # @raise [ValidationError, ServerError, ConnectionError]
      def add_replication_policy(policy)
        policy_obj = policy.is_a?(Models::ReplicationPolicy) ? policy : Models::ReplicationPolicy.new(policy)
        args = {
          "id" => policy_obj.id,
          "source_backend" => policy_obj.source_backend || "",
          "destination_backend" => policy_obj.destination_backend || "",
          "check_interval" => policy_obj.check_interval_seconds || 60,
          "enabled" => policy_obj.enabled || false
        }
        args["source_settings"] = policy_obj.source_settings if policy_obj.source_settings&.any?
        args["destination_settings"] = policy_obj.destination_settings if policy_obj.destination_settings&.any?
        args["source_prefix"] = policy_obj.source_prefix if policy_obj.source_prefix

        call_tool("objstore_add_replication_policy", args) do |result|
          { success: result.fetch("success", true), message: result["message"] }
        end
      end

      # Remove a replication policy.
      #
      # @param id [String] Policy ID
      #
      # @return [Hash] Response with :success and :message keys
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def remove_replication_policy(id)
        call_tool("objstore_remove_replication_policy", { "id" => id }) do |result|
          { success: result.fetch("success", true), message: result["message"] }
        end
      end

      # Get all replication policies.
      #
      # @return [Hash] Response with :policies array
      #
      # @raise [ServerError, ConnectionError]
      def get_replication_policies
        call_tool("objstore_list_replication_policies", {}) do |result|
          policies = extract_policies(result).map { |p| parse_replication_policy(p) }
          { policies: policies }
        end
      end

      # Get a specific replication policy.
      #
      # @param id [String] Policy ID
      #
      # @return [Hash] Response with :policy key
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def get_replication_policy(id)
        call_tool("objstore_get_replication_policy", { "id" => id }) do |result|
          { policy: parse_replication_policy(result) }
        end
      end

      # Trigger replication synchronization.
      #
      # @param policy_id [String, nil] Specific policy to trigger (nil for all)
      # @param parallel [Boolean] Ignored; included for API parity
      # @param worker_count [Integer] Ignored; included for API parity
      #
      # @return [Hash] Response with :success, :result, :message keys
      #
      # @raise [ServerError, ConnectionError]
      def trigger_replication(policy_id: nil, parallel: false, worker_count: 4)
        args = policy_id ? { "policy_id" => policy_id } : {}

        call_tool("objstore_trigger_replication", args) do |result|
          sync = result["result"] || {}
          {
            success: result.fetch("success", true),
            result: sync.transform_keys(&:to_sym),
            message: result["message"]
          }
        end
      end

      # Get replication status for a policy.
      #
      # @param id [String] Policy ID
      #
      # @return [Hash] Response with :success and :status keys
      #
      # @raise [NotFoundError, ServerError, ConnectionError]
      def get_replication_status(id)
        call_tool("objstore_get_replication_status", { "policy_id" => id }) do |result|
          {
            success: result.fetch("success", true),
            status: Models::ReplicationStatus.new(
              policy_id: result["policy_id"],
              source_backend: result["source_backend"],
              destination_backend: result["destination_backend"],
              enabled: result.fetch("enabled", false),
              total_objects_synced: result["total_objects_synced"] || 0,
              total_objects_deleted: result["total_objects_deleted"] || 0,
              total_bytes_synced: result["total_bytes_synced"] || 0,
              total_errors: result["total_errors"] || 0,
              average_sync_duration_ms: 0,
              sync_count: result["sync_count"] || 0
            )
          }
        end
      end

      # Close the persistent HTTP connection.
      #
      # The next request transparently reconnects. Safe to call multiple times
      # and before any connection has been established.
      #
      # @return [void]
      def close
        @http_mutex.synchronize { finish_http }
      end

      private

      # Build and send a JSON-RPC 2.0 "tools/call" request to the MCP endpoint,
      # parse the text result, and yield the parsed JSON hash to the block.
      #
      # @param tool_name [String] MCP tool name (e.g. "objstore_put")
      # @param arguments [Hash] Tool arguments
      #
      # @yieldparam result [Hash] Parsed JSON from result.content[0].text
      #
      # @return [Object] Whatever the block returns
      #
      # @raise [NotFoundError, ValidationError, ServerError, ConnectionError, TimeoutError]
      def call_tool(tool_name, arguments)
        payload = JSON.generate(
          jsonrpc: "2.0",
          method: "tools/call",
          params: { name: tool_name, arguments: arguments },
          id: next_id
        )

        response = http_post(payload)
        body = JSON.parse(response.body)

        if (err = body["error"])
          raise_rpc_error(err)
        end

        result_envelope = body.fetch("result", {})
        content = result_envelope.fetch("content", [])
        text = content.dig(0, "text") || "{}"

        parsed = begin
          JSON.parse(text)
        rescue JSON::ParserError
          { "data" => text }
        end

        yield parsed
      rescue ObjectStore::Error
        # Typed objstore errors (NotFound, Validation, Authentication,
        # Authorization, AlreadyExists, RateLimit, Server, Connection,
        # Timeout, Protocol) pass through unchanged.
        raise
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        raise ObjectStore::TimeoutError, e.message
      rescue JSON::ParserError => e
        raise ObjectStore::ProtocolError, "Invalid JSON from MCP server: #{e.message}"
      rescue StandardError => e
        raise ObjectStore::ConnectionError, e.message
      end

      # Build the persistent Net::HTTP connection. It is started lazily on
      # first use and reused across requests, avoiding a fresh TCP (and TLS)
      # handshake per call.
      def build_http
        http = Net::HTTP.new(@host, @port)
        http.use_ssl = @use_ssl
        http.verify_mode = OpenSSL::SSL::VERIFY_PEER if @use_ssl
        http.open_timeout = @timeout
        http.read_timeout = @timeout
        http.write_timeout = @timeout
        http
      end

      # Execute an HTTP POST to the MCP endpoint ("/") over the persistent
      # connection. When the server has closed an idle connection, the
      # connection is re-established and the request retried once.
      def http_post(body)
        request = Net::HTTP::Post.new(build_uri("/"))
        request["Content-Type"] = "application/json"
        request["Accept"] = "application/json"
        apply_auth_headers(request)

        response = @http_mutex.synchronize do
          retried = false
          begin
            @http.start unless @http.started?
            @http.request(request, body)
          rescue EOFError, Errno::ECONNRESET, Errno::EPIPE, IOError
            finish_http
            raise if retried

            retried = true
            retry
          end
        end

        handle_http_response(response)
      end

      # Close the underlying connection if it is open. Must be called while
      # holding @http_mutex (or from #close, which acquires it).
      def finish_http
        @http.finish if @http.started?
      rescue IOError
        # Already closed; nothing to release.
      end

      # Attach Authorization, X-Tenant-ID, and any extra headers.
      def apply_auth_headers(request)
        request["Authorization"] = "Bearer #{@token}" if @token
        request["X-Tenant-ID"] = @tenant_id if @tenant_id
        @extra_headers.each { |k, v| request[k.to_s] = v.to_s }
      end

      def build_uri(path)
        scheme = @use_ssl ? "https" : "http"
        URI("#{scheme}://#{@host}:#{@port}#{path}")
      end

      def handle_http_response(response)
        case response.code
        when "200", "201"
          response
        when "401"
          raise ObjectStore::AuthenticationError, "Unauthenticated"
        when "403"
          raise ObjectStore::AuthorizationError, "Forbidden"
        when "404"
          raise ObjectStore::NotFoundError, "Resource not found"
        when "400"
          body = safe_parse_body(response)
          raise ObjectStore::ValidationError, body["message"] || "Bad request"
        when "409"
          body = safe_parse_body(response)
          raise ObjectStore::AlreadyExistsError, body["message"] || "Already exists"
        when "429"
          body = safe_parse_body(response)
          raise ObjectStore::RateLimitError, body["message"] || "Rate limited"
        when /^5/
          body = safe_parse_body(response)
          raise ObjectStore::ServerError, body["message"] || "Server error"
        else
          raise ObjectStore::Error, "Unexpected response: #{response.code}"
        end
      end

      def safe_parse_body(response)
        return {} if response.body.nil? || response.body.empty?

        JSON.parse(response.body)
      rescue JSON::ParserError
        {}
      end

      # Decode the base64 "data" field of a tool result. Object data is
      # base64-encoded on the MCP transport; anything else is a protocol
      # violation and raises rather than being passed through silently.
      def decode_object_data(result)
        raw = result.fetch("data", "")
        return "" if raw.empty?

        Base64.strict_decode64(raw)
      rescue ArgumentError
        raise ObjectStore::ProtocolError, "invalid base64 data in MCP response"
      end

    end
  end
end

File: api/sdks/typescript/src/clients/mcp-client.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import {
  McpClientConfig,
  IObjectStoreClient,
  PutRequest,
  PutResponse,
  GetRequest,
  GetResponse,
  DeleteRequest,
  DeleteResponse,
  ListRequest,
  ListResponse,
  ExistsRequest,
  ExistsResponse,
  GetMetadataRequest,
  MetadataResponse,
  UpdateMetadataRequest,
  UpdateMetadataResponse,
  HealthRequest,
  HealthResponse,
  HealthStatus,
  ArchiveRequest,
  ArchiveResponse,
  AddPolicyRequest,
  AddPolicyResponse,
  RemovePolicyRequest,
  RemovePolicyResponse,
  GetPoliciesRequest,
  GetPoliciesResponse,
  ApplyPoliciesRequest,
  ApplyPoliciesResponse,
  AddReplicationPolicyRequest,
  AddReplicationPolicyResponse,
  RemoveReplicationPolicyRequest,
  RemoveReplicationPolicyResponse,
  GetReplicationPoliciesRequest,
  GetReplicationPoliciesResponse,
  GetReplicationPolicyRequest,
  GetReplicationPolicyResponse,
  TriggerReplicationRequest,
  TriggerReplicationResponse,
  GetReplicationStatusRequest,
  GetReplicationStatusResponse,
  replicationModeToString,
  stringToReplicationMode,
} from '../types';
import {
  validatePutRequest,
  validateGetRequest,
  validateDeleteRequest,
  validateListRequest,
  validateExistsRequest,
  validateGetMetadataRequest,
  validateUpdateMetadataRequest,
  validateArchiveRequest,
  validateAddPolicyRequest,
  validateRemovePolicyRequest,
  validateAddReplicationPolicyRequest,
  validateRemoveReplicationPolicyRequest,
  validateGetReplicationPolicyRequest,
  validateGetReplicationStatusRequest,
} from '../validation';
import { errorFromJsonRpc } from './jsonrpc';
import type { JsonRpcRequest, JsonRpcResponse } from './jsonrpc';

// JSON-RPC 2.0 envelope types shared with the Unix client (see ./jsonrpc),
// specialized for the MCP tools/call shape.
type McpJsonRpcRequest = JsonRpcRequest<{
  name: string;
  arguments: Record<string, unknown>;
}>;

type McpJsonRpcResponse = JsonRpcResponse<{
  content?: Array<{ type: string; text: string }>;
}>;

/**
 * McpClient communicates with go-objstore via the MCP HTTP transport
 * (pkg/server/mcp). Each operation calls HTTP POST to the base URL path "/"
 * with a JSON-RPC 2.0 envelope: method "tools/call", params { name, arguments }.
 * The tool result is a JSON string at result.content[0].text.
 */
export class McpClient implements IObjectStoreClient {
  private http: AxiosInstance;
  private nextId: number;

  constructor(config: McpClientConfig) {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      ...config.headers,
    };
    if (config.token) {
      headers['Authorization'] = `Bearer ${config.token}`;
    }
    if (config.tenantId) {
      headers['X-Tenant-ID'] = config.tenantId;
    }
    this.http = axios.create({
      baseURL: config.baseUrl,
      timeout: config.timeout ?? 30000,
      headers,
    });
    this.nextId = 1;
  }

  // callTool sends a tools/call JSON-RPC 2.0 request and returns the parsed
  // result object extracted from result.content[0].text.
  private async callTool(name: string, args: Record<string, unknown>): Promise<unknown> {
    const id = this.nextId++;
    const req: McpJsonRpcRequest = {
      jsonrpc: '2.0',
      method: 'tools/call',
      params: { name, arguments: args },
      id,
    };

    let resp: McpJsonRpcResponse;
    try {
      const axiosResp = await this.http.post<McpJsonRpcResponse>('/', req);
      resp = axiosResp.data;
    } catch (error) {
      throw this.handleAxiosError(error);
    }

    if (resp.error) {
      throw errorFromJsonRpc(resp.error, 'MCP');
    }

    const textContent = resp.result?.content?.[0]?.text;
    if (textContent === undefined) {
      return {};
    }

    try {
      return JSON.parse(textContent) as unknown;
    } catch {
      // Some tools return plain text; wrap it.
      return { text: textContent };
    }
  }

  async put(request: PutRequest): Promise<PutResponse> {
    validatePutRequest(request);
    const args: Record<string, unknown> = {
      key: request.key,
      data: request.data.toString('base64'),
    };
    if (request.metadata) {
      args.metadata = {
        content_type: request.metadata.contentType,
        content_encoding: request.metadata.contentEncoding,
        custom: request.metadata.custom,
      };
    }
    await this.callTool('objstore_put', args);
    return { success: true };
  }

  async get(request: GetRequest): Promise<GetResponse> {
    validateGetRequest(request);
    const result = (await this.callTool('objstore_get', { key: request.key })) as {
      data?: string;
      size?: number;
    };
    const raw = result.data ?? '';
    // The MCP server returns the object data base64-encoded in the tool result
    // JSON (matching how put encodes it); decode it back to the raw bytes.
    const data = Buffer.from(raw, 'base64');
    return { data, metadata: { size: result.size ?? data.length } };
  }

  async delete(request: DeleteRequest): Promise<DeleteResponse> {
    validateDeleteRequest(request);
    await this.callTool('objstore_delete', { key: request.key });
    return { success: true };
  }

  async list(request: ListRequest = {}): Promise<ListResponse> {
    validateListRequest(request);
    const args: Record<string, unknown> = {};
    if (request.prefix !== undefined) args.prefix = request.prefix;
    if (request.maxResults !== undefined) args.max_results = request.maxResults;
    if (request.continueFrom !== undefined) args.continue_from = request.continueFrom;

    const result = (await this.callTool('objstore_list', args)) as {
      keys?: string[];
      truncated?: boolean;
      next_token?: string;
    };
    return {
      objects: (result.keys ?? []).map((k) => ({ key: k })),
      commonPrefixes: [],
      nextToken: result.next_token,
      truncated: result.truncated ?? false,
    };
  }

  async exists(request: ExistsRequest): Promise<ExistsResponse> {
    validateExistsRequest(request);
    const result = (await this.callTool('objstore_exists', { key: request.key })) as {
      exists?: boolean;
    };
    return { exists: result.exists ?? false };
  }

  async getMetadata(request: GetMetadataRequest): Promise<MetadataResponse> {
    validateGetMetadataRequest(request);
    const result = (await this.callTool('objstore_get_metadata', { key: request.key })) as {
      size?: number;
      content_type?: string;
      content_encoding?: string;
      last_modified?: string;
      etag?: string;
      custom?: Record<string, string>;
    };
    return {
      success: true,
      metadata: {
        size: result.size,
        contentType: result.content_type,
        contentEncoding: result.content_encoding,
        lastModified: result.last_modified ? new Date(result.last_modified) : undefined,
        etag: result.etag,
        custom: result.custom,
      },
    };
  }

  async updateMetadata(request: UpdateMetadataRequest): Promise<UpdateMetadataResponse> {
    validateUpdateMetadataRequest(request);
    await this.callTool('objstore_update_metadata', {
      key: request.key,
      metadata: {
        content_type: request.metadata.contentType,
        content_encoding: request.metadata.contentEncoding,
        custom: request.metadata.custom,
      },
    });
    return { success: true };
  }

  async health(_request: HealthRequest = {}): Promise<HealthResponse> {
    const result = (await this.callTool('objstore_health', {})) as {
      status?: string;
    };
    const status =
      result.status === 'healthy'
        ? HealthStatus.SERVING
        : result.status === 'unhealthy'
        ? HealthStatus.NOT_SERVING
        : HealthStatus.UNKNOWN;
    return { status };
  }

  async archive(request: ArchiveRequest): Promise<ArchiveResponse> {
    validateArchiveRequest(request);
    await this.callTool('objstore_archive', {
      key: request.key,
      destination_type: request.destinationType,
      destination_settings: request.destinationSettings ?? {},
    });
    return { success: true };
  }

  async addPolicy(request: AddPolicyRequest): Promise<AddPolicyResponse> {
    validateAddPolicyRequest(request);
    await this.callTool('objstore_add_policy', {
      id: request.policy.id,
      prefix: request.policy.prefix,
      retention_seconds: request.policy.retentionSeconds,
      action: request.policy.action,
      destination_type: request.policy.destinationType,
      destination_settings: request.policy.destinationSettings,
    });
    return { success: true };
  }

  async removePolicy(request: RemovePolicyRequest): Promise<RemovePolicyResponse> {
    validateRemovePolicyRequest(request);
    await this.callTool('objstore_remove_policy', { id: request.id });
    return { success: true };
  }

  async getPolicies(request: GetPoliciesRequest = {}): Promise<GetPoliciesResponse> {
    const args: Record<string, unknown> = {};
    if (request.prefix !== undefined) args.prefix = request.prefix;
    const result = (await this.callTool('objstore_get_policies', args)) as {
      policies?: Array<{
        id: string;
        prefix?: string;
        action?: string;
        retention_seconds?: number;
        destination_type?: string;
        destination_settings?: Record<string, string>;
      }>;
    };
    return {
      success: true,
      policies: (result.policies ?? []).map((p) => ({
        id: p.id,
        prefix: p.prefix ?? '',
        action: p.action ?? '',
        retentionSeconds: p.retention_seconds ?? 0,
        destinationType: p.destination_type,
        destinationSettings: p.destination_settings,
      })),
    };
  }

  async applyPolicies(_request: ApplyPoliciesRequest = {}): Promise<ApplyPoliciesResponse> {
    const result = (await this.callTool('objstore_apply_policies', {})) as {
      policies_count?: number;
      objects_processed?: number;
    };
    return {
      success: true,
      policiesCount: result.policies_count ?? 0,
      objectsProcessed: result.objects_processed ?? 0,
    };
  }

  async addReplicationPolicy(
    request: AddReplicationPolicyRequest
  ): Promise<AddReplicationPolicyResponse> {
    validateAddReplicationPolicyRequest(request);
    await this.callTool('objstore_add_replication_policy', {
      id: request.policy.id,
      source_backend: request.policy.sourceBackend,
      source_settings: request.policy.sourceSettings,
      source_prefix: request.policy.sourcePrefix,
      destination_backend: request.policy.destinationBackend,
      destination_settings: request.policy.destinationSettings,
      check_interval: request.policy.checkIntervalSeconds,
      enabled: request.policy.enabled,
      replication_mode: replicationModeToString(request.policy.replicationMode),
      encryption: request.policy.encryption,
    });
    return { success: true };
  }

  async removeReplicationPolicy(
    request: RemoveReplicationPolicyRequest
  ): Promise<RemoveReplicationPolicyResponse> {
    validateRemoveReplicationPolicyRequest(request);
    await this.callTool('objstore_remove_replication_policy', { id: request.id });
    return { success: true };
  }

  async getReplicationPolicies(
    _request: GetReplicationPoliciesRequest = {}
  ): Promise<GetReplicationPoliciesResponse> {
    const result = (await this.callTool('objstore_list_replication_policies', {})) as {
      policies?: Array<{
        id: string;
        source_backend?: string;
        source_settings?: Record<string, string>;
        source_prefix?: string;
        destination_backend?: string;
        destination_settings?: Record<string, string>;
        check_interval?: number;
        last_sync_time?: string;
        enabled?: boolean;
        replication_mode?: string;
      }>;
    };
    return {
      policies: (result.policies ?? []).map((p) => ({
        id: p.id,
        sourceBackend: p.source_backend ?? '',
        sourceSettings: p.source_settings ?? {},
        sourcePrefix: p.source_prefix ?? '',
        destinationBackend: p.destination_backend ?? '',
        destinationSettings: p.destination_settings ?? {},
        checkIntervalSeconds: p.check_interval ?? 0,
        lastSyncTime: p.last_sync_time ? new Date(p.last_sync_time) : undefined,
        enabled: p.enabled ?? false,
        replicationMode: stringToReplicationMode(p.replication_mode ?? 'transparent'),
      })),
    };
  }

  async getReplicationPolicy(
    request: GetReplicationPolicyRequest
  ): Promise<GetReplicationPolicyResponse> {
    validateGetReplicationPolicyRequest(request);
    const result = (await this.callTool('objstore_get_replication_policy', {
      id: request.id,
    })) as {
      id?: string;
      source_backend?: string;
      source_settings?: Record<string, string>;
      source_prefix?: string;
      destination_backend?: string;
      destination_settings?: Record<string, string>;
      check_interval?: number;
      last_sync_time?: string;
      enabled?: boolean;
      replication_mode?: string;
    };
    return {
      policy: {
        id: result.id ?? request.id,
        sourceBackend: result.source_backend ?? '',
        sourceSettings: result.source_settings ?? {},
        sourcePrefix: result.source_prefix ?? '',
        destinationBackend: result.destination_backend ?? '',
        destinationSettings: result.destination_settings ?? {},
        checkIntervalSeconds: result.check_interval ?? 0,
        lastSyncTime: result.last_sync_time ? new Date(result.last_sync_time) : undefined,
        enabled: result.enabled ?? false,
        replicationMode: stringToReplicationMode(result.replication_mode ?? 'transparent'),
      },
    };
  }

  async triggerReplication(
    request: TriggerReplicationRequest
  ): Promise<TriggerReplicationResponse> {
    const args: Record<string, unknown> = {};
    if (request.policyId) args.policy_id = request.policyId;
    const result = (await this.callTool('objstore_trigger_replication', args)) as {
      result?: {
        policy_id?: string;
        synced?: number;
        deleted?: number;
        failed?: number;
        bytes_total?: number;
        duration?: string;
        errors?: string[];
      };
    };
    const r = result.result;
    return {
      success: true,
      result: r
        ? {
            policyId: r.policy_id ?? request.policyId ?? '',
            synced: r.synced ?? 0,
            deleted: r.deleted ?? 0,
            failed: r.failed ?? 0,
            bytesTotal: r.bytes_total ?? 0,
            durationMs: 0,
            errors: r.errors ?? [],
          }
        : undefined,
    };
  }

  async getReplicationStatus(
    request: GetReplicationStatusRequest
  ): Promise<GetReplicationStatusResponse> {
    validateGetReplicationStatusRequest(request);
    const result = (await this.callTool('objstore_get_replication_status', {
      policy_id: request.id,
    })) as {
      policy_id?: string;
      source_backend?: string;
      destination_backend?: string;
      enabled?: boolean;
      total_objects_synced?: number;
      total_objects_deleted?: number;
      total_bytes_synced?: number;
      total_errors?: number;
      last_sync_time?: string;
      average_sync_duration?: string;
      sync_count?: number;
    };
    return {
      success: true,
      status: {
        policyId: result.policy_id ?? request.id,
        sourceBackend: result.source_backend ?? '',
        destinationBackend: result.destination_backend ?? '',
        enabled: result.enabled ?? true,
        totalObjectsSynced: result.total_objects_synced ?? 0,
        totalObjectsDeleted: result.total_objects_deleted ?? 0,
        totalBytesSynced: result.total_bytes_synced ?? 0,
        totalErrors: result.total_errors ?? 0,
        lastSyncTime: result.last_sync_time ? new Date(result.last_sync_time) : undefined,
        averageSyncDurationMs: 0,
        syncCount: result.sync_count ?? 0,
      },
    };
  }

  async close(): Promise<void> {
    // Nothing to close for the MCP HTTP client.
    return Promise.resolve();
  }

  private handleAxiosError(error: unknown): Error {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      const message =
        (axiosError.response?.data as Record<string, unknown>)?.message ||
        (axiosError.response?.data as Record<string, unknown>)?.error ||
        axiosError.message;
      const status = axiosError.response?.status ?? 500;
      return new Error(`MCP HTTP error (${status}): ${message}`);
    }
    return error instanceof Error ? error : new Error(String(error));
  }
}
